### PR TITLE
[zprezto-update] Add convenience function to update zprezto

### PR DESCRIPTION
**Pull request by samcv**
This function checks if there is any update to zprezto, and if so
will pull in the changes. It will not attempt a pull unless
it is fastforwardable. It also makes sure the user is on the master
branch before attempting.

### CODE REVIEW

**Comment by Yoshyn** [comment]
Pretty cool PR. 
Just one question : I have modification on theses files : 
  * runcoms/zpreztorc (adding module, & co ...)
  * runcoms/zshrc        (export EDITOR & VISUAL...

So, if i run this it fail due to : 
```
error: cannot pull with rebase: You have unstaged changes.
error: please commit or stash them.
```

Manually, i stash it and reapply after.
I think that a lot people modify theses files. Do you plan to manage it?

**Comment by diraol** [comment]
Following up @Yoshyn comment, what could have been done here is something like what is done on `spf13-vim` distro.
We could have a `zprestorc.local` and a `zshrc.local` files that are imported on the end of the `zprestorc` and `zshrc` files. Doing so, the `.local` won't be versioned and then the users personalized configurations wouldn't get in conflict with the main config files, making this updates easier.
Does it make sense?

Also, what do you think about printing, as output from the update command, the messages from the commits that were pulled, so the user can know what was updated? Does it make sense, at least with an optional argument?

**Comment by Yoshyn** [comment]
Well, initially i though about someting like that : 
```
printf "Try to stash before pulling..."
local zprezto_stash="zprezto-update-`date +%s`"
git stash save "$(echo $zprezto_stash)" --include-untracked
if git pull --ff-only; then
    git submodule update --recursive
    git stash pop
else
    printf "Unable to fast-forward the changes. You can fix this by running 'cd %s' and then 'git pull' to manually pull and possibly merge in changes\n"
fi
```

But it's just unmaintainable.  
After reading @diraol, i'm completely convinced. A local setup not versioned will be awesome

File: init.zsh [hunk 1]
```diff
@@ -17,6 +17,29 @@ if ! autoload -Uz is-at-least || ! is-at-least "$min_zsh_version"; then
 fi
 unset min_zsh_version
 
+# zprezto convenience updater
+function zprezto-update {
+    local OWD="$(pwd)"
+    cd -- "${ZPREZTODIR}"
```
Comment (belak) [inline init.zsh]: `pushd` and `popd` would probably be better for this. Then we don't need to track `OWD`. Also note that we should use `-q` to ensure the chpwd functions are not run.

File: init.zsh [hunk 2]
```diff
@@ -17,6 +17,29 @@ if ! autoload -Uz is-at-least || ! is-at-least "$min_zsh_version"; then
...
+    local orig_branch="$(git symbolic-ref HEAD 2>/dev/null | cut -d '/' -f 3)"
+    if [[ "$orig_branch" == "master" ]]; then
+        git fetch
+        if [[ "$(git status -uno | grep behind)" ]]; then
+            printf "There is an update availible. Trying to pull"
+            if git pull --ff-only; then
+                printf "Syncing submodules\n"
+                git submodule update --recursive
+            else
+                printf "Unable to fast-forward the changes. You can fix this by running 'cd %s' and then 'git pull' to manually pull and possibly merge in changes\n"
```
Comment (belak) [inline init.zsh]: Looks like this is missing the param for the `%s`

File: init.zsh [hunk 3]
```diff
@@ -17,6 +17,29 @@ if ! autoload -Uz is-at-least || ! is-at-least "$min_zsh_version"; then
...
 unset min_zsh_version
 
+# zprezto convenience updater
+function zprezto-update {
+    local OWD="$(pwd)"
+    cd -- "${ZPREZTODIR}"
+    local orig_branch="$(git symbolic-ref HEAD 2>/dev/null | cut -d '/' -f 3)"
+    if [[ "$orig_branch" == "master" ]]; then
+        git fetch
+        if [[ "$(git status -uno | grep behind)" ]]; then
```
Comment (belak) [inline init.zsh]: This check seems a bit error prone. Maybe something like what's mentioned here would be better: https://stackoverflow.com/a/3278427 (on a side note, I had no idea `@` even existed in git)

Comment (samcv) [inline init.zsh]: Took that suggestion and implemented it in my latest commit.

File: init.zsh (hunk 1 above)
Comment (samcv) [inline init.zsh]: Fixed this in the latest commit.

File: init.zsh (hunk 2 above)
Comment (samcv) [inline init.zsh]: Fixed.

**Comment by samcv** [comment]
Fixed @belak's suggestions in my latest commit and cleaned up a few things regarding the error handling. As stated in the commit description.

File: init.zsh [hunk 4]
```diff
@@ -17,6 +17,45 @@ if ! autoload -Uz is-at-least || ! is-at-least "$min_zsh_version"; then
...
+function zprezto-update {
+   function cannot-fast-forward {
+       local STATUS="$1"
+       [[ "${STATUS}" ]] && printf "%s\n" "${STATUS}"
+       printf "Unable to fast-forward the changes. You can fix this by "
+       printf "running\ncd '%s' and then\n'git pull' " "${ZPREZTODIR}"
+       printf "to manually pull and possibly merge in changes\n"
+   }
+   pushd -q
+   cd -- "${ZPREZTODIR}"
```
Comment (belak) [inline init.zsh]: This would be better as `pushd -q -- "${ZPREZTODIR}"` so cd won't cause chpwd hooks to run (cd also has a similar option, but they can be combined, so it makes sense to)

File: init.zsh [hunk 5]
```diff
@@ -17,6 +17,45 @@ if ! autoload -Uz is-at-least || ! is-at-least "$min_zsh_version"; then
 fi
 unset min_zsh_version
 
+# zprezto convenience updater
+function zprezto-update {
+   function cannot-fast-forward {
+       local STATUS="$1"
+       [[ "${STATUS}" ]] && printf "%s\n" "${STATUS}"
```
Comment (belak) [inline init.zsh]: Should this be `[[ -n "${STATUS}" ]]`?

File: init.zsh [hunk 6]
```diff
@@ -17,6 +17,45 @@ if ! autoload -Uz is-at-least || ! is-at-least "$min_zsh_version"; then
...
+       elif [[ $REMOTE == $BASE ]]; then
+           cannot-fast-forward "Commits in master that aren't in upstream."
+       else
+           cannot-fast-forward "Upstream and local have diverged."
+       fi
+   else
+       printf "zprezto install at '%s' is not on the master branch " "${ZPREZTODIR}"
+       printf "(you're on '%s')\nUnable to automatically update.\n" "${orig_branch}"
+   fi
+   popd -q
```
Comment (belak) [inline init.zsh]: It would be nice to have proper return codes if this fails. `popd` could be moved to an atexit function (`trap 'popd -q' EXIT`) which would make it easier to return without having to copy the popd everywhere.

File: init.zsh (hunk 5 above)
Comment (samcv) [inline init.zsh]: `[[ -n "${STATUS}" ]]` is the same thing as `[[ "${STATUS}" ]]`.

Should I explicitly add the -n? (this works without -n even in posix shell by the way)

Comment (belak) [inline init.zsh]: This is fine, I think... because we control the value of STATUS. If we didn't the -n would make sense.

File: init.zsh (hunk 6 above)
Comment (samcv) [inline init.zsh]: Okay, I had to make a change and not run pushd -q earlier on to make popd -q work inside the trap. Otherwise it would take me back to the wrong folder.

File: runcoms/zprofile [hunk 7]
```diff
@@ -65,12 +65,9 @@ fi
 #
 # Temporary Files
 #
-#
-if [[ -z "$TMPDIR" ]]; then
```
Comment (belak) [inline runcoms/zprofile]: This change seems unrelated...

File: init.zsh [hunk 8]
```diff
@@ -17,6 +17,51 @@ if ! autoload -Uz is-at-least || ! is-at-least "$min_zsh_version"; then
...
+# zprezto convenience updater
+function zprezto-update {
+  function cannot-fast-forward {
+    local STATUS="$1"
+    [[ "${STATUS}" ]] && printf "%s\n" "${STATUS}"
+    printf "Unable to fast-forward the changes. You can fix this by "
+    printf "running\ncd '%s' and then\n'git pull' " "${ZPREZTODIR}"
+    printf "to manually pull and possibly merge in changes\n"
+  }
+  cd -- "${ZPREZTODIR}" || return 7
```
Comment (belak) [inline init.zsh]: This will still trigger chpwd callbacks (which we don't want to run). There's an option to cd which won't trigger them... Also, I'm a bit confused why pushd didn't work...

**Comment by samcv** [comment]
I've added the `-q` option to `cd`. Instead of using pushd and trying to popd on exit, I changed it to use `( )` around the function which keeps the contents in the function outside the external scope and avoids the issue of having to change back. Since the `cd` happens inside a different scope.

**Comment by belak** [comment]
Alright, a subshell works. I'd like to see a comment mentioning that (because it's easy to miss that `()` was used rather than `{}`.

Once that's done and a small blurb is added to the readme (or somewhere relevant) I think this is ready.

**Comment by samcv** [comment]
@belak I've added information to the README and added a comment to `init.zsh` about it being in a subshell.

File: init.zsh [hunk 9]
```diff
@@ -17,6 +17,52 @@ if ! autoload -Uz is-at-least || ! is-at-least "$min_zsh_version"; then
...
+    local STATUS="$1"
+    [[ "${STATUS}" ]] && printf "%s\n" "${STATUS}"
+    printf "Unable to fast-forward the changes. You can fix this by "
+    printf "running\ncd '%s' and then\n'git pull' " "${ZPREZTODIR}"
+    printf "to manually pull and possibly merge in changes\n"
+  }
+  cd -q -- "${ZPREZTODIR}" || return 7
+  local orig_branch="$(git symbolic-ref HEAD 2>/dev/null | cut -d '/' -f 3)"
+  if [[ "$orig_branch" == "master" ]]; then
+    git fetch
```
Comment (belak) [inline init.zsh]: Sorry, forgot about this one. We should probably return if git fetch fails.

Comment (samcv) [inline init.zsh]: Done.

**Comment by belak** [comment]
Merged! Thanks for your work on this!

**Comment by samcv** [comment]
We can probably have an enhancement based on @Yoshyn's suggestion to possibly stash things and then reapply but I'd think it'd be best to keep it simple for at least a little while and make sure there's no issues with things as they have been merged in currently. Since at least we know it won't make any mistakes if it aborts in case of conflicts.

**Comment by belak** [comment]
Yep! I figured, since this is new and not used by many people it's alright to have a few non-critical bugs... and this PR has had enough feedback on it. :)
